icefoganalytics/tourism-activity-statistics-tracking
# Issue 1: Add Build Visitor Counting Interface

# Overview

Some visitor centres use door counters, and at the end of the day, get an overall number of visitors, rather than doing input incrementally. They want to first enter the overall total, then distribute the visitors into various categories base on -- I'm guessing -- paper records that they have been keeping through the day. They need to see the difference (unallocated) to know they got the middle numbers correct.

- Fixes #1

# Implementation

Bulk edit mode makes use of the existing API endpoint by setting the appropriate "delta" for the given location. I opted not to clear the database values, and instead make use of the existing architecture for better traceability and reversibility.

Visitors of "Unknown" origin can never be set directly, only as a result of setting the "total" visitors, and having uncategorized visitors left over. This vastly reduces complexity and enhances data integrity.

I've used heavy JS validation to prevent the user from entering invalid input, but this validation is silent. This validation is all on the front-end, and does not include and form validation warnings. Invalid data simply does not persist in the UI.

When a User reduces the number of total visitors, this wipes all visitor category totals, and moves all values back to the "unknown" category. An alternative solution would be to lock edit of the "Total visitors" field if any data exists in any of the other fields; however, this seems like it would have the same effect, and be much more work for the User.

# Questions/Concerns
- [ ] Should I duplicate/refactor the `/record-stats` endpoint such that full replace is possible? This vastly increases complexity, with uncertain benefit.
- [ ] Should I share the data entry card header between the two edit modes? This vastly increases complexity, but would result in shared visitor centre and date values. Currently implementation resets when "edit" mode is toggled.
- [ ] The validation logic is pretty complicated, and completely untested. Is it worth configuring a test suite just for this feature?
- [ ] Should I wrap the build edit mode feature in a "form" tag and use form validation errors for accessibility reasons?

# Test Instructions

1. Boot the database via `docker compose --env-file ./.env.development --file docker-compose.dev.yaml up --build`
2. Boot the API back-end via `cd src/api && npm run start`
3. Boot the Vue front-end via `cd src/web && npm run start`
4. Log in using your "Government of Yukon online services account" to http://localhost:8080
5. Open your database editor and connect to the database with the credential from your `db/sapassword.env` file.
e.g.
```
user => sa
password => DifficultPassword123!
```
6. Make your user an admin by going to the `user#is_admin` -> `1` via your database editor.
7. Go back to the app at http://localhost:8080, and log out and log back.
8. Go to the Administraion section of the top right kebab menu.
![image](https://github.com/icefoganalytics/tourism-activity-statistics-tracking/assets/23045206/18ee32d5-3af5-491d-a2b0-bf90c1b2ecb2)
9. Click "Users".
10. Click your user to edit it.
11. Add permissions for all sites.
12. Add all sites to "sites to manage".
13. Click save.
14. Click home, or go to http://localhost:8080/dashboard.
15. Select a Visitor center and a date.
16. Click to toggle the "Bulk Input" mode.
> Note that this currently resets the select visitor centre and date, this should probably get fixed? This could be fixed a number of ways, the simplest being moving the load/select visitor centre date to a "created" instead of a "mounted" hook, though that might create other problems.
17. Enter a daily visitor total such as `100`.
18. Note that the number of "Unknown" visitors increase by the same amount.
19. Enter some visitors in the various columns.
20. Note that you cannot overflow the max daily total visitors.
21. Note that the "Unknown" visitors value decreases as you distribute amounts to other locations.
22. Click save.
23. You will be redirected back to the "incremental" data entry mode with the new values present.
24. Switch back to "bulk" entry mode, note that the values carry back over.
25. Reduce the number of total visitors. Note that this wipes all visitors totals and moves all values back to the "unknown" category.
26. Click "save" again and check that the values persist.
